### PR TITLE
Add shutter curve functionality

### PR DESCRIPTION
**Pull request by fedormatantsev**
- Fixed linear curve inverse sampling;
- Implemented parametric curve-based shutter;

constant
curve_linear
curve_nonlinear

### CODE REVIEW

File: src/appleseed/renderer/modeling/camera/camera.cpp [hunk 1]
```diff
@@ -311,44 +368,218 @@ double Camera::extract_near_z() const
...
     }
     else
     {
         // Shutter is fully opened.
-        return sample / m_shutter_pdf_max_height + m_normalized_open_end_time_half;
+        return m_shutter_curve_linear_fully_opened_multiplier * sample + m_shutter_open_end_time / 2.0f;
+    }
+}
+
+// Nonlinear curve consists of two bezier curves and constant patch between them
```
Comment (dictoon) [inline src/appleseed/renderer/modeling/camera/camera.cpp]: *bezier* -> *Bezier*

File: src/appleseed/renderer/modeling/camera/camera.cpp [hunk 2]
```diff
@@ -311,44 +368,218 @@ double Camera::extract_near_z() const
...
+float Camera::map_to_shutter_curve_impl_nonlinear(const float sample) const
```
Comment (dictoon) [inline src/appleseed/renderer/modeling/camera/camera.cpp]: Maybe `_bezier` is clearer than `_nonlinear`?

File: src/appleseed/renderer/modeling/camera/camera.cpp [hunk 3]
```diff
@@ -311,44 +368,218 @@ double Camera::extract_near_z() const
...
+                                          sample,
+                                          0.0f,
+                                          m_shutter_curve_nonlinear_control_points_normalized[0],
+                                          m_shutter_curve_nonlinear_control_points_normalized[2],
+                                          m_normalized_open_end_time,
+                                          *this);
+    }
+    else if (sample <= m_inverse_cdf_close_point)
+    {
+        return  (sample - m_inverse_cdf_open_point) /
```
Comment (dictoon) [inline src/appleseed/renderer/modeling/camera/camera.cpp]: Can we fold the division into `m_shutter_curve_nonlinear_normalization_factor`?

File: src/appleseed/renderer/modeling/camera/camera.cpp [hunk 4]
```diff
@@ -311,44 +368,218 @@ double Camera::extract_near_z() const
...
+
+float Camera::get_shutter_curve_nonlinear_normalization_factor() const
+{
+    return 1.0f / m_shutter_curve_nonlinear_close_cdf.evaluate(1.0f);
+}
+
+float Camera::map_to_shutter_curve(const float sample) const
+{
+    assert(sample < 1.0f);
+    if (m_use_nonlinear_shutter_curve)
```
Comment (dictoon) [inline src/appleseed/renderer/modeling/camera/camera.cpp]: You're free to omit the curly braces for single line `if`s. Here using the ternary operator leads to an even more compact notation:

    return m_use_nonlinear_shutter_curve
        ? map_to_shutter_curve_impl_nonlinear(sample)
        : map_to_shutter_curve_impl_linear(sample);

File: src/appleseed/renderer/modeling/camera/camera.cpp [hunk 5]
```diff
@@ -502,6 +732,13 @@ DictionaryArray CameraFactory::get_input_metadata()
             .insert("use", "optional")
             .insert("default", "1.0"));
 
+    metadata.push_back(
+    Dictionary()
```
Comment (dictoon) [inline src/appleseed/renderer/modeling/camera/camera.cpp]: Indent.

File: src/appleseed/renderer/modeling/camera/camera.h [hunk 6]
```diff
@@ -38,6 +38,7 @@
 // appleseed.foundation headers.
 #include "foundation/math/dual.h"
 #include "foundation/math/vector.h"
+#include "foundation/math/polynomial.h"
```
Comment (dictoon) [inline src/appleseed/renderer/modeling/camera/camera.h]: Order alphabetically.

File: src/appleseed/renderer/modeling/camera/camera.h [hunk 7]
```diff
@@ -157,21 +158,39 @@ class APPLESEED_DLLSYMBOL Camera
...
     float               m_normalized_close_start_time;
-    float               m_open_linear_curve_slope;
-    float               m_close_linear_curve_slope;
-    float               m_shutter_pdf_max_height;
     float               m_inverse_cdf_open_point;
     float               m_inverse_cdf_close_point;
     bool                m_motion_blur_enabled;
 
+    // Linear shutter related parameters
+    float m_shutter_curve_linear_open_multiplier;
```
Comment (dictoon) [inline src/appleseed/renderer/modeling/camera/camera.h]: Align members vertically.

File: src/appleseed/renderer/modeling/camera/camera.h [hunk 8]
```diff
@@ -198,16 +217,38 @@ class APPLESEED_DLLSYMBOL Camera
     // Map a sample using inverse of CDF calculated from camera shutter graph. Used in initialize_ray().
     float map_to_shutter_curve(const float sample) const;
 
+        // Check shutter times and emit warnings if needed.
```
Comment (dictoon) [inline src/appleseed/renderer/modeling/camera/camera.h]: Fix indentation.

File: src/appleseed/renderer/modeling/camera/camera.h [hunk 9]
```diff
@@ -198,16 +217,38 @@ class APPLESEED_DLLSYMBOL Camera
...
+    void check_shutter_times_for_consistency() const;
+
+    // Map a sample to a composition of two lines and a constant. Used in map_to_shutter_curve().
+    float map_to_shutter_curve_impl_linear(const float sample) const;
+
+    // Map a sample to a composition of two bezier curves and a constant. Used in map_to_shutter_curve().
```
Comment (dictoon) [inline src/appleseed/renderer/modeling/camera/camera.h]: *Bezier*

File: src/appleseed/renderer/utility/paramarray.h [hunk 10]
```diff
@@ -109,7 +109,7 @@ class APPLESEED_DLLSYMBOL ParamArray
     template <typename T>
     T get_required(
         const char*             name,
-        const T&                default_value) const;
+        const T&                default_value = T()) const;
```
Comment (dictoon) [inline src/appleseed/renderer/utility/paramarray.h]: Agree with that, but it deserves a separate commit.

**Comment by dictoon** [comment]
Also, it would be nice to add test scenes to the test suite (in `sandbox/tests/test scenes/`) to check that now and in the future that this feature works as intended.

File: src/appleseed/foundation/math/polynomial.h [hunk 11]
```diff
@@ -0,0 +1,112 @@
...
+{
+
+//
+// Polynomial Class
+//
+// Represents N-th order polynomial with the given
+// coefficients (in monomial basis) of type T and parameter of type P
+//
+template <typename T, typename P, int N>
+class Polynomial : public Vector<T, N+1>
```
[the remark this answers is not in the document]
Comment (fedormatantsev) [reply src/appleseed/foundation/math/polynomial.h]: Well, as an example there is Bernstein polynomial basis and monomial polynomial basis, actually polynomials make linear space and we can set summation and multiplication operators for them that would work; I suppose we can even retrieve isomorphism for vector space and polynomial space, which actually means that those guys are equivalent from algebra's point of view :)

File: src/appleseed/foundation/math/polynomial.h [hunk 12]
```diff
@@ -0,0 +1,112 @@
...
+    //  This function yields into p0 + p * (p1 + p * (p2 + p * (...+ p * pN))...))
+    //
+
+    size_t i = N;
+    T value = (*this)[i];
+
+    do
+    {
+        --i;
+        value = param * value + (*this)[i];
```
[the remark this answers is not in the document]
Comment (fedormatantsev) [reply src/appleseed/foundation/math/polynomial.h]: My bad! Fixed it.

File: src/appleseed/foundation/math/polynomial.h [hunk 13]
```diff
@@ -0,0 +1,112 @@
...
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+//
+
+#ifndef APPLESEED_FOUNDATION_MATH_POLYNOMIAL_H
+#define APPLESEED_FOUNDATION_MATH_POLYNOMIAL_H
+
+#include "foundation/math/vector.h"
```
[the remark this answers is not in the document]
Comment (fedormatantsev) [reply src/appleseed/foundation/math/polynomial.h]: size_t is used in Vector.h, so cstddef is redundant here.
Added comment.

File: src/appleseed/foundation/math/polynomial.h [hunk 14]
```diff
@@ -0,0 +1,112 @@
...
+// coefficients (in monomial basis) of type T and parameter of type P
+//
+template <typename T, typename P, int N>
+class Polynomial : public Vector<T, N+1>
+{
+  public:
+    T evaluate(const P& param) const;
+    T evaluate_derivative(const P& param) const;
+
+    // Multiplies every coefficient of the polinomial by value
```
[the remark this answers is not in the document]
Comment (fedormatantsev) [reply src/appleseed/foundation/math/polynomial.h]: Fixed.

File: src/appleseed/foundation/math/polynomial.h [hunk 15]
```diff
@@ -0,0 +1,112 @@
...
+    while (i > 0);
+
+    return value;
+}
+
+template <typename T, typename P, int N>
+T Polynomial<T, P, N>::evaluate_derivative(const P& param) const
+{
+    //
+    //  This function yields into p1 + p * (2 * p2 + p * (...+ N * p * pN))...))
```
[the remark this answers is not in the document]
Comment (fedormatantsev) [reply src/appleseed/foundation/math/polynomial.h]: Fixed.

File: src/appleseed/foundation/math/polynomial.h [hunk 16]
```diff
@@ -0,0 +1,112 @@
...
+
+    size_t i = N;
+    T value = N * (*this)[i];
+
+    do
+    {
+        --i;
+        value = param * value + (P)i * (*this)[i];
+    }
+    while(i > 1);
```
[the remark this answers is not in the document]
Comment (fedormatantsev) [reply src/appleseed/foundation/math/polynomial.h]: Removed this line.

File: src/appleseed/foundation/math/polynomial.h [hunk 17]
```diff
@@ -0,0 +1,112 @@
...
+    {
+        result[i] = (*this)[i] * value;
+    }
+
+    return result;
+}
+
+} // namespace foundation
+
+#endif
```
[the remark this answers is not in the document]
Comment (fedormatantsev) [reply src/appleseed/foundation/math/polynomial.h]: Fixed.

File: src/appleseed/foundation/math/polynomial.h [hunk 18]
```diff
@@ -0,0 +1,112 @@
...
+//
+template <typename T, typename P, int N>
+class Polynomial : public Vector<T, N+1>
+{
+  public:
+    T evaluate(const P& param) const;
+    T evaluate_derivative(const P& param) const;
+
+    // Multiplies every coefficient of the polinomial by value
+    Polynomial<T, P, N> operator*(const T& value) const;
```
[the remark this answers is not in the document]
Comment (fedormatantsev) [reply src/appleseed/foundation/math/polynomial.h]: Missed that; Will remove this code.

File: src/appleseed/renderer/modeling/camera/camera.cpp [hunk 19]
```diff
@@ -38,6 +38,8 @@
 
 // appleseed.foundation headers.
 #include "foundation/math/scalar.h"
+#include "foundation/math/bezier.h"
```
[the remark this answers is not in the document]
Comment (fedormatantsev) [reply src/appleseed/renderer/modeling/camera/camera.cpp]: Fixed.

File: src/appleseed/renderer/modeling/camera/camera.cpp [hunk 20]
```diff
@@ -86,20 +89,74 @@ bool Camera::on_render_begin(
...
     {
-        m_normalized_open_end_time = inverse_lerp(m_shutter_open_time, m_shutter_close_time, m_shutter_open_end_time);
-        m_normalized_open_end_time_half = m_normalized_open_end_time / 2.0f;
-        m_normalized_close_start_time = inverse_lerp(m_shutter_open_time, m_shutter_close_time, m_shutter_close_start_time);
-        m_shutter_pdf_max_height = 2.0f / (1.0f + m_normalized_close_start_time - m_normalized_open_end_time);
-        m_open_linear_curve_slope = m_shutter_pdf_max_height / m_normalized_open_end_time;
-        m_close_linear_curve_slope = -m_shutter_pdf_max_height / (1.0f - m_normalized_close_start_time);
-        m_inverse_cdf_open_point = m_shutter_pdf_max_height * m_normalized_open_end_time_half;
-        m_inverse_cdf_close_point = m_shutter_pdf_max_height * (m_normalized_close_start_time - m_normalized_open_end_time_half);
+        // Initialize curve shutter parameters
```
[the remark this answers is not in the document]
Comment (fedormatantsev) [reply src/appleseed/renderer/modeling/camera/camera.cpp]: Fixed.

File: src/appleseed/renderer/modeling/camera/camera.cpp [hunk 21]
```diff
@@ -86,20 +89,74 @@ bool Camera::on_render_begin(
...
+        m_use_nonlinear_shutter_curve = true;
+
+        m_shutter_curve_nonlinear_control_points_normalized =
+                m_params.get_required<ShutterCurveControlPoints>("shutter_curve_control_points");
```
[the remark this answers is not in the document]
Comment (fedormatantsev) [reply src/appleseed/renderer/modeling/camera/camera.cpp]: Fixed.

File: src/appleseed/renderer/modeling/camera/camera.cpp [hunk 22]
```diff
@@ -86,20 +89,74 @@ bool Camera::on_render_begin(
...
+
+        // Normilize time points.
```
[the remark this answers is not in the document]
Comment (fedormatantsev) [reply src/appleseed/renderer/modeling/camera/camera.cpp]: Fixed.

File: src/appleseed/renderer/modeling/camera/camera.cpp [hunk 23]
```diff
@@ -86,20 +89,74 @@ bool Camera::on_render_begin(
...
+        for (size_t i = 0; i<ShutterCurveControlPoints::Dimension; i+=2)
```
[the remark this answers is not in the document]
Comment (fedormatantsev) [reply src/appleseed/renderer/modeling/camera/camera.cpp]: Fixed.

File: src/appleseed/renderer/modeling/camera/camera.cpp [hunk 24]
```diff
@@ -86,20 +89,74 @@ bool Camera::on_render_begin(
...
+        {
+            m_shutter_curve_nonlinear_control_points_normalized[i] =
+                    inverse_lerp(m_shutter_open_time,
```
[the remark this answers is not in the document]
Comment (fedormatantsev) [reply src/appleseed/renderer/modeling/camera/camera.cpp]: Done.

File: src/appleseed/renderer/modeling/camera/camera.cpp [hunk 25]
```diff
@@ -86,20 +89,74 @@ bool Camera::on_render_begin(
...
+                                 m_shutter_close_time,
+                                 m_shutter_curve_nonlinear_control_points_normalized[i]);
+        }
+
+        // Use normalized time, so inverse sampling will end up with normilized time points.
```
[the remark this answers is not in the document]
Comment (fedormatantsev) [reply src/appleseed/renderer/modeling/camera/camera.cpp]: Done.

File: src/appleseed/renderer/modeling/camera/camera.cpp [hunk 26]
```diff
@@ -311,44 +368,218 @@ double Camera::extract_near_z() const
 
 namespace
 {
+//
```
[the remark this answers is not in the document]
Comment (fedormatantsev) [reply src/appleseed/renderer/modeling/camera/camera.cpp]: Okay, thanks for explanations!
Fixed.

File: src/appleseed/renderer/modeling/camera/camera.cpp [hunk 27]
```diff
@@ -311,44 +368,218 @@ double Camera::extract_near_z() const
...
+inline float inverse_sample_curve_patch(
+    const foundation::Polynomial<float, float, 6>& curve,
+    const float value,
+    const float t0,
+    const float t1,
+    const float t2,
+    const float t3,
+    const Camera& camera)
+{
+    const float eps = 1e-6;
```
[the remark this answers is not in the document]
Comment (fedormatantsev) [reply src/appleseed/renderer/modeling/camera/camera.cpp]: Wow! How did I miss this guy?!
Fixed.

File: src/appleseed/renderer/modeling/camera/camera.cpp [hunk 28]
```diff
@@ -311,44 +368,218 @@ double Camera::extract_near_z() const
 
 namespace
 {
+//
+// Used in map_to_shutter_curve_impl_nonlinear()
+//
+inline float inverse_sample_curve_patch(
+    const foundation::Polynomial<float, float, 6>& curve,
```
[the remark this answers is not in the document]
Comment (fedormatantsev) [reply src/appleseed/renderer/modeling/camera/camera.cpp]: Removed all the foundation:: prefixes

File: src/appleseed/renderer/modeling/camera/camera.cpp [hunk 29]
```diff
@@ -311,44 +368,218 @@ double Camera::extract_near_z() const
...
+    {
+        // make sure neuton method didn't push root slightly out of bounds
+        root = foundation::clamp(root, a, max_param);
+
+        return foundation::evaluate_bezier3(t0, t1, t2, t3, root);
+    }
+
     //
-    // Shutter curves. Used in Camera::map_to_shutter_curve().
+    // Fallback to bisection.
```
[the remark this answers is not in the document]
Comment (fedormatantsev) [reply src/appleseed/renderer/modeling/camera/camera.cpp]: Yes, I've got some samples that do not converge (they are close to p=0 or p=1); Basically, clamping will eliminate those; Fallback bisection will let me sleep normally though;

File: src/appleseed/renderer/modeling/camera/camera.cpp [hunk 30]
```diff
@@ -311,44 +368,218 @@ double Camera::extract_near_z() const
...
+        a,
+        b,
+        eps,
+        10,
+        root
+    );
+
+    if (success)
+    {
+        // make sure neuton method didn't push root slightly out of bounds
```
[the remark this answers is not in the document]
Comment (fedormatantsev) [reply src/appleseed/renderer/modeling/camera/camera.cpp]: Fixed.

File: src/appleseed/renderer/modeling/camera/camera.cpp [hunk 31]
```diff
@@ -311,44 +368,218 @@ double Camera::extract_near_z() const
...
-            constant = 0.0f;
-        }
-        else
-        {
-            // Shutter is closing.
-            constant = 1.0f - (a / 2.0f) - b;
-        }
-        return (sqrt(2.0f * a * (x - constant) + b * b) - b) / a;
+        assert(false && "Cannot inverse sample shutter curve CDF");
+        RENDERER_LOG_ERROR("Cannot inverse sample shutter curve CDF of camera \"%s\"", camera.get_path().c_str());
```
[the remark this answers is not in the document]
Comment (fedormatantsev) [reply src/appleseed/renderer/modeling/camera/camera.cpp]: I stopped rendering when it was 10 million samples; So probability of getting this assertion is less then 1/10000000;

File: src/appleseed/renderer/modeling/camera/camera.cpp (hunk 1 above)
Comment (fedormatantsev) [reply src/appleseed/renderer/modeling/camera/camera.cpp]: Fixed.

File: src/appleseed/foundation/math/polynomial.h (hunk 11 above)
[the remark this answers is not in the document]
Comment (dictoon) [reply src/appleseed/foundation/math/polynomial.h]: It's true that a polynomial is simply a point in the ring of polynomials so in a sense that makes sense. Let's keep it like this.

File: src/appleseed/renderer/modeling/camera/camera.cpp (hunk 2 above)
Comment (fedormatantsev) [reply src/appleseed/renderer/modeling/camera/camera.cpp]: I wanted to name it _bezier, but finally we've got something that is not a Bezier curve itself, so nonlinear curve is more accurate name.
Should be discussed further though.

File: src/appleseed/renderer/modeling/camera/camera.cpp (hunk 3 above)
Comment (fedormatantsev) [reply src/appleseed/renderer/modeling/camera/camera.cpp]: You mean use multiplication by 1/m_shutter_curve_normalization_factor?

File: src/appleseed/foundation/math/polynomial.h (hunk 13 above)
[the remark this answers is not in the document]
Comment (dictoon) [reply src/appleseed/foundation/math/polynomial.h]: While this is true, we follow the IWYU ("Include What You Use") rule. This means that every file (.h and .cpp) must include all header files it relies on, even if those header files are already included by others.

The _only_ exception is that .cpp files must not repeat includes already present in their "interface" header file, i.e. `foo.cpp` must not repeat includes from `foo.h`.

The IWYU rule, or at least the term, was popularized by Google via their iwyu tool: https://include-what-you-use.org/

Sometimes we run this tool on appleseed's code base and fix some includes, but there are way too many violations at the moment to do a full pass.

File: src/appleseed/renderer/modeling/camera/camera.cpp (hunk 4 above)
Comment (fedormatantsev) [reply src/appleseed/renderer/modeling/camera/camera.cpp]: Cool;
In cqg we don't omit braces even for one-statement-long body; Habits :)
Fixed.

File: src/appleseed/renderer/modeling/camera/camera.cpp (hunk 5 above)
Comment (fedormatantsev) [reply src/appleseed/renderer/modeling/camera/camera.cpp]: Done.

File: src/appleseed/renderer/modeling/camera/camera.h (hunk 6 above)
Comment (fedormatantsev) [reply src/appleseed/renderer/modeling/camera/camera.h]: Done.

File: src/appleseed/renderer/modeling/camera/camera.cpp (hunk 31 above)
[the remark this answers is not in the document]
Comment (dictoon) [reply src/appleseed/renderer/modeling/camera/camera.cpp]: Cool, then if we're guaranteed to find a solution, I would get rid of the message and just keep the assertion.

By the way, a common trick for message assertions is this:

    assert(!"Cannot inverse sample shutter curve CDF");

File: src/appleseed/renderer/modeling/camera/camera.h (hunk 7 above)
Comment (fedormatantsev) [reply src/appleseed/renderer/modeling/camera/camera.h]: Can I align fields by blocks (with different indentation for each block)?

File: src/appleseed/renderer/modeling/camera/camera.h (hunk 8 above)
Comment (fedormatantsev) [reply src/appleseed/renderer/modeling/camera/camera.h]: Fixed.

File: src/appleseed/renderer/modeling/camera/camera.cpp (hunk 2 above)
Comment (dictoon) [reply src/appleseed/renderer/modeling/camera/camera.cpp]: It's not a Bézier curve if you consider the whole curve, but it's still a combination of linear and Bézier segments. If we were to add support for other curves (B-Spline?) then the `_nonlinear` suffix would be a problem.

File: src/appleseed/renderer/modeling/camera/camera.cpp (hunk 3 above)
Comment (dictoon) [reply src/appleseed/renderer/modeling/camera/camera.cpp]: Yeah, I would change `m_shutter_curve_nonlinear_normalization_factor` to include the 1/ part.

File: src/appleseed/renderer/modeling/camera/camera.cpp (hunk 4 above)
Comment (dictoon) [reply src/appleseed/renderer/modeling/camera/camera.cpp]: Sure, it's completely correct to keep braces, at my work we also decided to force using them. In appleseed you're free to decide. I just wanted to point out a more compact notation, that's all.

File: src/appleseed/renderer/modeling/camera/camera.h (hunk 7 above)
Comment (dictoon) [reply src/appleseed/renderer/modeling/camera/camera.h]: Yes, sure. Aligning members is more of an art form than a science, there's no rule that always work, and there's certainly a level of subjectivity to what looks more pleasing.

File: src/appleseed/renderer/modeling/camera/camera.h (hunk 9 above)
Comment (fedormatantsev) [reply src/appleseed/renderer/modeling/camera/camera.h]: Fixed.

File: src/appleseed/renderer/utility/paramarray.h (hunk 10 above)
Comment (fedormatantsev) [reply src/appleseed/renderer/utility/paramarray.h]: I'll make a separate commit for that during final rebase

File: src/appleseed/foundation/math/polynomial.h (hunk 13 above)
[the remark this answers is not in the document]
Comment (fedormatantsev) [reply src/appleseed/foundation/math/polynomial.h]: Okay, I'll fix it. Thanks for explanation!

File: src/appleseed/foundation/math/polynomial.h (hunk 18 above)
[the remark this answers is not in the document]
Comment (fedormatantsev) [reply src/appleseed/foundation/math/polynomial.h]: Remove Polynomial class, use free functions instead.

File: src/appleseed/renderer/modeling/camera/camera.cpp (hunk 31 above)
[the remark this answers is not in the document]
Comment (fedormatantsev) [reply src/appleseed/renderer/modeling/camera/camera.cpp]: Okay, I'll remove the message.

File: src/appleseed/renderer/modeling/camera/camera.cpp (hunk 2 above)
Comment (fedormatantsev) [reply src/appleseed/renderer/modeling/camera/camera.cpp]: Let it be bezier then.

File: src/appleseed/renderer/modeling/camera/camera.cpp (hunk 3 above)
Comment (fedormatantsev) [reply src/appleseed/renderer/modeling/camera/camera.cpp]: Actually, it has that 1/ part. We are INVERSE sampling, so this is denormalization:) That is why we have division here.

**Comment by fedormatantsev** [comment]
TODO: remove clamping in inverse_sample_curve_patch() as Newton method won't go out of bounds now.

**Comment by dictoon** [comment]
Thanks!

It looks the unit tests fail, see Travis log.

**Comment by fedormatantsev** [comment]
TODO: use bisection step instead of clamping in find_root_newton() (utilize computed f(root) and recompute delta) this should work better. This will allow totally wipe out fallback bisection step in inverse sampling code.

File: src/appleseed/foundation/math/root.h [hunk 32]
```diff
@@ -157,7 +157,11 @@ bool find_root_newton(
         const T delta = f(root) / d(root);
         root -= delta;
 
-        if (std::abs(delta) <= eps)
+        if ((root < a) || (root > b))
```
[the remark this answers is not in the document]
Comment (fedormatantsev) [reply src/appleseed/foundation/math/root.h]: Well, we could always clamp the root, yes. Good suggestion! Missd that.

File: src/appleseed/renderer/utility/transformsequence.cpp [hunk 33]
```diff
@@ -561,20 +561,23 @@ AABB3d TransformSequence::compute_motion_segment_bbox(
         const TrajectoryX tx(sx, sy, corner2d);
         const TrajectoryY ty(sx, sy, corner2d);
 
+        const double a = angle > 0.0 ? 0.0 : angle;
```
[the remark this answers is not in the document]
Comment (fedormatantsev) [reply src/appleseed/renderer/utility/transformsequence.cpp]: Yes, totally agree, will change that part

File: src/appleseed/foundation/math/polynomial.h [hunk 34]
```diff
@@ -0,0 +1,88 @@
...
+{
+
+//
+// Evaluates a polynomial represented by a vector of coefficients (in monomial basis).
+// Yields p0 + p * (p1 + p * (p2 + p * (...+ p * p{N-1})...)).
+//
+template <typename T, typename P, size_t N>
+T evaluate_polynomial(const Vector<T, N>& coefficients, const P& param)
+{
+    assert(N > 0);
```
[the remark this answers is not in the document]
Comment (fedormatantsev) [reply src/appleseed/foundation/math/polynomial.h]: fixed

File: src/appleseed/foundation/math/polynomial.h [hunk 35]
```diff
@@ -0,0 +1,88 @@
...
+//
+template <typename T, typename P, size_t N>
+T evaluate_polynomial_derivative(const Vector<T, N>& coefficients, const P& param)
+{
+    assert(N > 0);
+
+    // Derivative of a constant is zero.
+    if (N == 1)
+    {
+        return (T)0;
```
[the remark this answers is not in the document]
Comment (fedormatantsev) [reply src/appleseed/foundation/math/polynomial.h]: Done

**Comment by dictoon** [comment]
Thanks for the excellent work!
